daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Relation.scala

```diff
@@ -35,6 +35,12 @@ object Relation {
 
       mutMap.toMap
     }
+
+    def flatten[A, B](relation: Relation[A, B]): Iterator[(A, B)] =
+      for {
+        key <- relation.keysIterator
+        value <- relation(key)
```

Comment:
```suggestion
        (key, values) <- relation.iterator
        value <- values
```
@S11001001 What about this?

Reply:
Don't do that, it adds a `withFilter` call.

Reply:
Yes, even though it looks innocent. You can't even write `kvs: (A, Set[B]) <-` without it adding a `withFilter`; the only safe things are variable bindings and `_`.

Reply:
I thought `withFilter` is only applied if a guard is in the `for`-comprehension? Can you explain, please?

Reply:
A pattern counts as a guard (passing iff the pattern matches), and anything counts as a pattern for this purpose except the two exceptions I mentioned above, inducing a `withFilter` even if they appear irrefutable.

Reply:
Sure, the only issue with the two lines above is that they don't compile. Can I call `._2` on `kvs` or are there caveats on that?

Reply:
I'll assume there are no further caveats.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/4922/commits/2c2c06646ac38fdd9939fd925f74256f54b46f95